internal/controller/agenttoken_controller.go

```diff
@@ -0,0 +1,421 @@
+			if err != nil && err != tfc.ErrResourceNotFound {
+				t.log.Error(err, "Reconcile Agent Token", "msg", fmt.Sprintf("failed to remove token %q", id))
+				return err
+			}
+			s := &corev1.Secret{}
+			nn := types.NamespacedName{
+				Namespace: t.instance.Namespace,
+				Name:      t.instance.Spec.SecretName,
+			}
+			t.log.Info("Reconcile Agent Token", "msg", fmt.Sprintf("remove key=%q from in secret=%q namespace=%q", id, nn.Name, nn.Namespace))
+			if err := r.Client.Get(ctx, nn, s); err != nil {
+				if apierrors.IsNotFound(err) {
+					t.instance.Status.AgentTokens = slice.RemoveFromSlice(t.instance.Status.AgentTokens, i)
+					return nil
+				}
+				t.log.Error(err, "Reconcile Agent Token", "msg", fmt.Sprintf("failed to get secret=%q namespace=%q", nn.Name, nn.Namespace))
+				return err
+			}
+			//
+			patch := client.MergeFrom(s.DeepCopy())
+			delete(s.Data, token.Name)
+			if err := r.Client.Patch(ctx, s, patch); err != nil {
+				t.log.Error(err, "Reconcile Agent Token", "msg", fmt.Sprintf("unable to remove key=%q in secret=%q namespace=%q", token.Name, nn.Name, nn.Namespace))
+				return err
+			}
+			t.log.Info("Reconcile Agent Token", "msg", fmt.Sprintf("successfully removed key=%q in secret=%q namespace=%q", token.Name, nn.Name, nn.Namespace))
+			t.instance.Status.AgentTokens = slice.RemoveFromSlice(t.instance.Status.AgentTokens, i)
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (r *AgentTokenReconciler) reconcileToken(ctx context.Context, t *agentTokenInstance) error {
+	t.log.Info("Reconcile Agent Token", "msg", "reconciling agent token")
+
+	// verify whether the Kubernetes object has been marked as deleted and if so delete the tokens
+	if isDeletionCandidate(&t.instance, agentTokenFinalizer) {
+		t.log.Info("Reconcile Agent Token", "msg", "object marked as deleted, need to delete tokens first")
+		r.Recorder.Event(&t.instance, corev1.EventTypeNormal, "ReconcileAgentToken", "Object marked as deleted, need to delete tokens first")
+		return r.deleteAgentToken(ctx, t)
+	}
+
+	tokens, err := r.listAgentTokens(ctx, t)
+	if err != nil {
+		return err
+	}
+
+	statusTokens := make(map[string]string, len(t.instance.Status.AgentTokens))
+	for _, token := range t.instance.Status.AgentTokens {
+		statusTokens[token.Name] = token.ID
+	}
+
+	// Clean up.
+	for _, token := range t.instance.Spec.AgentTokens {
+		if tokenID, ok := statusTokens[token.Name]; ok {
+			delete(statusTokens, token.Name)
+			if _, ok := tokens[tokenID]; ok {
+				delete(tokens, tokenID)
+				continue
+			}
+			if err := r.removeToken(ctx, t, tokenID); err != nil {
+				return err
+			}
+		}
+		if err := r.createToken(ctx, t, token.Name); err != nil {
+			return err
+		}
+	}
+	for _, id := range statusTokens {
+		if err := r.removeToken(ctx, t, id); err != nil {
+			return err
+		}
+	}
+
+	switch t.instance.Spec.ManagementPolicy {
+	case appv1alpha2.AgentTokenManagementPolicyMerge:
+		// This remains no-op.
```

Comment:
Because this case is a no-op we don't need to include it in the switch and then wecan instead just formulate the case below as an if statement.